src/stack/Stack.h

```diff
@@ -0,0 +1,295 @@
+struct IsBleLayerConfiguration : std::is_base_of<BleLayerConfiguration, T>
+{
+};
+
+class DefaultBleLayer : BleLayerConfiguration
+{
+public:
+    CHIP_ERROR Init(const StackParameters & parameters) { return CHIP_NO_ERROR; }
+    CHIP_ERROR Shutdown() { return CHIP_NO_ERROR; }
+
+    Ble::BleLayer * Get() { return nullptr; }
+};
+
+/* ========================== Transport Configuration ========================== */
+struct TransportConfiguration
+{
+};
+template <typename T>
+struct IsTransportConfiguration : std::is_base_of<TransportConfiguration, T>
+{
+};
+
+class DefaultTransport : TransportConfiguration
+{
+public:
+    using transport = TransportMgr<
+#if INET_CONFIG_ENABLE_IPV4
+        Transport::UDP, /* IPv4 */
+#endif
+        Transport::UDP>;
+
+    CHIP_ERROR Init(const StackParameters & parameters, Inet::InetLayer & inetLayer, Ble::BleLayer * bleLayer)
+    {
+        return mTransportManager.Init(
+#if INET_CONFIG_ENABLE_IPV4
+            Transport::UdpListenParameters(&inetLayer)
+                .SetAddressType(Inet::kIPAddressType_IPv4)
+                .SetListenPort(parameters.GetListenPort()),
+#endif
+            Transport::UdpListenParameters(&inetLayer)
+                .SetAddressType(Inet::kIPAddressType_IPv6)
+                .SetListenPort(parameters.GetListenPort()));
+    }
+
+    CHIP_ERROR Shutdown()
+    {
+        mTransportManager.Close();
+        return CHIP_NO_ERROR;
+    }
+
+    TransportMgrBase & Get() { return mTransportManager; }
+
+private:
+    transport mTransportManager;
+};
+
+/* ========================== Stack ========================== */
+
+/**
+ * @brief
+ *   Chip Stack is an interface to manage all chip modules for applications.
+ *
+ *   This class should not be used inside CHIP SDK.
+ */
+template <typename... Configurations>
+class Stack
+{
+private:
+    using SystemLayerConfig =
+        typename first_if_any_or_default<IsSystemLayerConfiguration, DefaultSystemLayer, Configurations...>::type;
+    using InetLayerConfig = typename first_if_any_or_default<IsInetLayerConfiguration, DefaultInetLayer, Configurations...>::type;
+    using StorageConfig   = typename first_if_any_or_default<IsStorageConfiguration, DefaultStorage, Configurations...>::type;
+    using BleLayerConfig  = typename first_if_any_or_default<IsBleLayerConfiguration, DefaultBleLayer, Configurations...>::type;
+    using TransportConfig = typename first_if_any_or_default<IsTransportConfiguration, DefaultTransport, Configurations...>::type;
+
+public:
+    Stack(NodeId localDeviceId) : mLocalDeviceId(localDeviceId) {}
+    virtual ~Stack() {}
+
+    // TODO(#6931): CHIP initialization should be posted to chip thread, in order to avoid racing problems
+    CHIP_ERROR Init(const StackParameters & parameters)
+    {
+        ReturnErrorOnFailure(Platform::MemoryInit());
+#if CONFIG_DEVICE_LAYER
+        // Initialize the CHIP stack.
+        ReturnErrorOnFailure(DeviceLayer::PlatformMgr().InitChipStack());
```

Comment:
FYI: For some reason, currently `ThreadStackMgr().InitThreadStack();` is also needed for Thread devices, so it either should be started here or incorporated into `InitChipStack()` when `CHIP_DEVICE_CONFIG_ENABLE_THREAD`

Reply:
I saw that `ThreadStackMgr().InitThreadStack()` is called after `InitChipStack()` for platforms with thread support. Maybe we should separate hardware initialization and chip stack initialization, and leave `InitChipStack()` outside chip stack.

Reply:
I'm not sure why ThreadStackManager can't be initialized by InitChipStack (which initializes other platform components). Maybe it's because ThreadStackManager can post events to the CHIP thread and the intention was not to initialize ThreadStackManager before the CHIP thread is started. Anyway, in case you plan to move the CHIP thread creation to the CHIP stack it would make sense to initialize ThreadStackManager in the comon code, too.